### PR TITLE
Flush 'add using' caches when we encounter low memory conditions.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

File: src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs [hunk 1]
```diff
@@ -76,9 +76,22 @@ public MetadataInfo(DateTime timeStamp, SymbolTreeInfo info, HashSet<ProjectId>
...
+
             return new IncrementalAnalyzer(_projectToInfo, _metadataPathToInfo);
         }
 
+        private void OnCacheFlushRequested(object sender, EventArgs e)
+        {
+            // If we hear about low memory conditions, flush our caches.  This will degrade the 
+            // experience a bit (as we will no longer offer to Add-Using for p2p refs/metadata),
+            // but will be better than OOM'ing.  These caches will be regenerated in the future
+            // when the incremental analyzer reanalyzers the projects in teh workspace.
```
Comment (rchande) [inline src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs]: "reanalyzes"
"the"

**Comment by rchande** [comment]
:+1:

**Comment by KevinH-MS** [comment]
Looks right to me (modulo typos) :+1:

**Comment by davkean** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/3293/
// Retest reason: https://github.com/dotnet/roslyn/issues/8305

**Comment by davkean** [comment]
@tannergooding Notice that "tst" job failed: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_tst_unit64/121/ failed

File: src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs [hunk 2]
```diff
@@ -76,9 +76,25 @@ public MetadataInfo(DateTime timeStamp, SymbolTreeInfo info, HashSet<ProjectId>
 
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
         {
+            var cacheService = workspace.Services.GetService<IWorkspaceCacheService>();
```
Comment (jaredpar) [inline src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs]: In what circumstances would this be `null`?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs]: When the workspace doesn't expose such a service.  An example of this is the workspace we have in EditorTestApp.

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/win/tst/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_tst_unit64/121/
// Retest reason: timeout

**Pull request merged**